File: WinPaletter/GUI/RetroElements/Input/TextBoxR.vb
Imports System.ComponentModel
Imports System.Drawing.Drawing2D


Namespace UI.Retro
    <Description("Retro TextBox with Windows 9x style")> <DefaultEvent("TextChanged")> Public Class TextBoxR : Inherits Control

        Sub New()
            SetStyle(ControlStyles.AllPaintingInWmPaint Or ControlStyles.UserPaint Or
                 ControlStyles.ResizeRedraw Or ControlStyles.OptimizedDoubleBuffer Or ControlStyles.SupportsTransparentBackColor, True)
            DoubleBuffered = True
            ForeColor = Color.Black
            BackColor = Color.White
            TB = New Windows.Forms.TextBox With {.Visible = True}
            Font = New Font("Microsoft Sans Serif", 8)
            TB.Text = Text
            TB.ForeColor = Color.White
            TB.MaxLength = _MaxLength
            TB.Multiline = _Multiline
            TB.ReadOnly = _ReadOnly
            TB.UseSystemPasswordChar = _UseSystemPasswordChar
            TB.BorderStyle = BorderStyle.None
            TB.Location = New Point(1, 0)
            TB.Width = Width - 1
            _Style = RoundingStyle.Normal
            TB.Cursor = Cursors.IBeam

            If _Multiline Then
                TB.Height = Height - 8
            Else
                Height = TB.Height + 8
            End If

            AddHandler TB.TextChanged, AddressOf OnBaseTextChanged
            AddHandler TB.KeyDown, AddressOf OnBaseKeyDown
        End Sub

#Region "Variables"
        Private State As MouseState = MouseState.None
        Private WithEvents TB As Windows.Forms.TextBox
        Enum MouseState As Byte
            None = 0
            Over = 1
            Down = 2
            Block = 3
        End Enum

#End Region

#Region "Properties"
#Region "TextBox Properties"

        Private _TextAlign As HorizontalAlignment = HorizontalAlignment.Left

        Public Enum RoundingStyle
            Normal
            Rounded
        End Enum

        Private _Style As RoundingStyle

        <Category("Options")>
        Property Style() As RoundingStyle
            Get
                Return _Style
            End Get
            Set(ByVal value As RoundingStyle)
                _Style = value
                If TB IsNot Nothing Then
                    TB.TextAlign = CType(value, HorizontalAlignment)
                End If
            End Set
        End Property

        Private _BorderStyle As BorderStyle

        <Category("Options")>
        Public Property BorderStyle() As BorderStyle
            Get
                Return _BorderStyle
            End Get
            Set(ByVal value As BorderStyle)
                _BorderStyle = value
                Invalidate()
            End Set
        End Property

        <Category("Options")>
        Property TextAlign() As HorizontalAlignment
            Get
                Return _TextAlign
            End Get
            Set(ByVal value As HorizontalAlignment)
                _TextAlign = value
                If TB IsNot Nothing Then
                    TB.TextAlign = value
                End If
            End Set
        End Property

        Private _MaxLength As Integer = 32767

        <Category("Options")>
        Property MaxLength() As Integer
            Get
                Return _MaxLength
            End Get
            Set(ByVal value As Integer)
                _MaxLength = value
                If TB IsNot Nothing Then
                    TB.MaxLength = value
                End If
            End Set
        End Property

        Private _ReadOnly As Boolean

        <Category("Options")>
        Property [ReadOnly]() As Boolean
            Get
                Return _ReadOnly
            End Get
            Set(ByVal value As Boolean)
                _ReadOnly = value
                If TB IsNot Nothing Then
                    TB.ReadOnly = value
                End If
            End Set
        End Property

        Private _UseSystemPasswordChar As Boolean

        <Category("Options")>
        Property UseSystemPasswordChar() As Boolean
            Get
                Return _UseSystemPasswordChar
            End Get
            Set(ByVal value As Boolean)
                _UseSystemPasswordChar = value
                If TB IsNot Nothing Then
                    TB.UseSystemPasswordChar = value
                End If
            End Set
        End Property

        Private _Multiline As Boolean

        <Category("Options")>
        Property Multiline() As Boolean
            Get
                Return _Multiline
            End Get
            Set(ByVal value As Boolean)
                _Multiline = value
                If TB IsNot Nothing Then
                    TB.Multiline = value

                    If value Then
                        TB.Height = Height - 8
                    Else
                        Height = TB.Height + 8
                    End If

                End If
            End Set
        End Property

        <Category("Options")>
        Overrides Property Text As String
            Get
                Return MyBase.Text
            End Get
            Set(ByVal value As String)
                MyBase.Text = value
                If TB IsNot Nothing Then
                    TB.Text = value
                End If
            End Set
        End Property

        <Category("Options")>
        Overrides Property Font As Font
            Get
                Return MyBase.Font
            End Get
            Set(ByVal value As Font)
                MyBase.Font = value
                If TB IsNot Nothing Then
                    TB.Font = value
                    TB.Location = New Point(4, 4)
                    TB.Width = Width - 8
                    If Not _Multiline Then
                        Height = TB.Height + 10
                    End If
                End If
            End Set
        End Property

        Protected Overrides Sub OnCreateControl()
            MyBase.OnCreateControl()
            If Not Controls.Contains(TB) Then
                Controls.Add(TB)
            End If
        End Sub

        Private Sub OnBaseTextChanged(ByVal s As Object, ByVal e As EventArgs)
            Text = TB.Text
        End Sub

        Private Sub OnBaseKeyDown(ByVal s As Object, ByVal e As KeyEventArgs)
            If e.Control AndAlso e.KeyCode = Keys.A Then
                TB.SelectAll()
                e.SuppressKeyPress = True
            End If
            If e.Control AndAlso e.KeyCode = Keys.C Then
                TB.Copy()
                e.SuppressKeyPress = True
            End If
            Invalidate()
        End Sub

        Protected Overrides Sub OnResize(ByVal e As EventArgs)
            TB.Location = New Point(4, 4)
            TB.Width = Width - 8

            If _Multiline Then
                TB.Height = Height - 8
            Else
                Height = TB.Height + 10
            End If

            MyBase.OnResize(e)
        End Sub

#End Region
#End Region
        Public Property ButtonShadow As Color = Color.FromArgb(128, 128, 128)
        Public Property ButtonDkShadow As Color = Color.Black
        Public Property ButtonHilight As Color = Color.White
        Public Property ButtonLight As Color = Color.FromArgb(192, 192, 192)
#Region "Colors"
        Private ReadOnly _BaseColor As Color = BackColor
        Private ReadOnly _TextColor As Color = ForeColor
#End Region

#Region "Events"
        Protected Overrides Sub OnMouseDown(e As MouseEventArgs)
            MyBase.OnMouseDown(e)
            State = MouseState.Down : Invalidate() : TB.Focus()
        End Sub

        Protected Overrides Sub OnMouseUp(e As MouseEventArgs)
            MyBase.OnMouseUp(e)
            State = MouseState.Over : TB.Focus() : Invalidate()
        End Sub

        Protected Overrides Sub OnMouseEnter(ByVal e As EventArgs)
            MyBase.OnMouseEnter(e)
            State = MouseState.Over : Invalidate()
        End Sub

        Protected Overrides Sub OnMouseLeave(ByVal e As EventArgs)
            MyBase.OnMouseLeave(e)
            State = MouseState.None : Invalidate()
        End Sub
#End Region
        Protected Overrides Sub OnPaint(e As PaintEventArgs)
            Dim B As New Bitmap(Width, Height)
            Dim G As Graphics = Graphics.FromImage(B)
            G = Graphics.FromImage(B)
            DoubleBuffered = True
            G.SmoothingMode = SmoothingMode.HighSpeed
            G.TextRenderingHint = My.RenderingHint

            MyBase.OnPaint(e)
            TB.ForeColor = ForeColor

            '################################################################################# Customizer
            Dim CheckRect As New Rectangle(0, 0, Width - 1, Height - 1)
            '#################################################################################

            G.Clear(BackColor)

            With CheckRect
                G.DrawLine(New Pen(ButtonShadow), New Point(.X, .Y), New Point(.Width - 1, .Y))
                G.DrawLine(New Pen(ButtonShadow), New Point(.X, .Y), New Point(.X, .Height - 1))
                G.DrawLine(New Pen(ButtonDkShadow), New Point(.X, .Y) + New Point(1, 1), New Point(.Width - 2, .Y + 1))
                G.DrawLine(New Pen(ButtonDkShadow), New Point(.X, .Y) + New Point(1, 1), New Point(.X + 1, .Height - 2))
                G.DrawLine(New Pen(ButtonLight), New Point(.Width - 1, 1), New Point(.Width - 1, .Height - 1))
                G.DrawLine(New Pen(ButtonLight), New Point(1, .Height - 1), New Point(.Width - 1, .Height - 1))
                G.DrawLine(New Pen(ButtonHilight), New Point(.Width, .X), New Point(.Width, .Height))
                G.DrawLine(New Pen(ButtonHilight), New Point(.X, .Height), New Point(.Width, .Height))
            End With

            G.DrawString(TB.Text, Font, New SolidBrush(ForeColor), New Point(2, 4))

            G.Dispose()
            e.Graphics.DrawImageUnscaled(B, 0, 0)
            B.Dispose()
        End Sub

        Private Sub TB_MouseDown(sender As Object, e As MouseEventArgs) Handles TB.MouseDown
            State = MouseState.Down
            Invalidate()
        End Sub

        Private Sub TB_MouseEnter(sender As Object, e As EventArgs) Handles TB.MouseEnter
            State = MouseState.Over : Invalidate()
        End Sub

        Private Sub TB_MouseLeave(sender As Object, e As EventArgs) Handles TB.MouseLeave
            State = MouseState.None : Invalidate()
        End Sub

        Private Sub TB_LostFocus(sender As Object, e As EventArgs) Handles TB.LostFocus
            State = MouseState.None : Invalidate()
        End Sub

        Private Sub RetroTextBox_BackColorChanged(sender As Object, e As EventArgs) Handles Me.BackColorChanged
            Try : If TB IsNot Nothing Then TB.BackColor = BackColor
            Catch : End Try
        End Sub
    End Class
End Namespace

File: WinPaletter/GUI/RetroElements/Input/TextBoxR.cs
using System;
using System.ComponentModel;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Drawing.Text;
using System.Runtime.CompilerServices;
using System.Windows.Forms;

namespace WinPaletter.UI.Retro
{
    /// <summary>
    /// A control that wraps a standard <see cref="TextBox"/> and paints a Windows 9x
    /// sunken 3D border around it. The inner TextBox handles all text input natively.
    /// </summary>
    [Description("Retro TextBox with Windows 9x style")]
    [DefaultEvent("TextChanged")]
    public class TextBoxR : Control
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="TextBoxR"/> class.
        /// </summary>
        public TextBoxR()
        {
            SetStyle(ControlStyles.OptimizedDoubleBuffer | ControlStyles.AllPaintingInWmPaint | ControlStyles.UserPaint | ControlStyles.ResizeRedraw | ControlStyles.SupportsTransparentBackColor, true);

            ForeColor = SystemColors.WindowText;
            BackColor = SystemColors.Window;
            Font = new Font("Microsoft Sans Serif", 8f);

            // Inner TextBox — borderless, positioned inside the painted border.
            TB = new()
            {
                Visible = true,
                Font = new Font("Microsoft Sans Serif", 8f),
                ForeColor = SystemColors.WindowText,
                BackColor = SystemColors.Window,
                MaxLength = _maxLength,
                Multiline = _multiline,
                ReadOnly = _readOnly,
                UseSystemPasswordChar = _useSystemPasswordChar,
                BorderStyle = BorderStyle.None,
                Cursor = Cursors.IBeam
            };

            TB.TextChanged += OnInnerTextChanged;
            TB.KeyDown += OnInnerKeyDown;

            UpdateInnerBounds();
        }

        #region Inner TextBox

        private TextBox _tb;

        /// <summary>
        /// The borderless inner <see cref="TextBox"/> that handles text input.
        /// Event subscriptions are managed through the setter to avoid leaks.
        /// </summary>
        private TextBox TB
        {
            [MethodImpl(MethodImplOptions.Synchronized)]
            get => _tb;

            [MethodImpl(MethodImplOptions.Synchronized)]
            set
            {
                if (_tb != null)
                {
                    _tb.MouseDown -= TB_MouseDown;
                    _tb.MouseEnter -= TB_MouseEnter;
                    _tb.MouseLeave -= TB_MouseLeave;
                    _tb.LostFocus -= TB_LostFocus;
                }

                _tb = value;

                if (_tb != null)
                {
                    _tb.MouseDown += TB_MouseDown;
                    _tb.MouseEnter += TB_MouseEnter;
                    _tb.MouseLeave += TB_MouseLeave;
                    _tb.LostFocus += TB_LostFocus;
                }
            }
        }

        /// <summary>
        /// Repositions and resizes the inner TextBox to sit inside the 4px border padding.
        /// In single-line mode the outer control height is driven by the inner TextBox height.
        /// </summary>
        private void UpdateInnerBounds()
        {
            if (TB == null) return;

            TB.Location = new Point(4, 4);
            TB.Width = Math.Max(0, Width - 8);

            if (_multiline)
            {
                TB.Height = Math.Max(0, Height - 8);
            }
            else
            {
                Height = TB.Height + 10;
            }
        }

        #endregion

        #region Cached GDI resources

        // Cached pens — rebuilt only when the corresponding color property changes.
        private Pen _penShadow;
        private Pen _penDkShadow;
        private Pen _penHilight;
        private Pen _penLight;

        // Cached border segment endpoints — rebuilt on resize.
        private PointF[] _dkShadowSeg0, _dkShadowSeg1;
        private PointF[] _shadowSeg0, _shadowSeg1;
        private PointF[] _hilightSeg0, _hilightSeg1;
        private PointF[] _lightSeg0, _lightSeg1;

        private static void ReplacePen(ref Pen pen, Color color)
        {
            pen?.Dispose();
            pen = new Pen(color);
        }

        /// <summary>
        /// Recomputes all cached border segment endpoints from the current control size.
        /// Win9x sunken border layout (outer→inner, top-left then bottom-right):
        ///   DkShadow  outer top + left
        ///   Shadow    inner top + left
        ///   Hilight   inner bottom + right
        ///   Light     outer bottom + right
        /// </summary>
        private void RebuildGeometry()
        {
            float w = Width - 1f;
            float h = Height - 1f;

            _dkShadowSeg0 = [new(0, 0), new(w, 0)];
            _dkShadowSeg1 = [new(0, 0), new(0, h)];
            _shadowSeg0 = [new(1, 1), new(w - 1f, 1)];
            _shadowSeg1 = [new(1, 1), new(1, h - 1f)];
            _hilightSeg0 = [new(w - 1f, 1), new(w - 1f, h - 1f)];
            _hilightSeg1 = [new(1, h - 1f), new(w - 1f, h - 1f)];
            _lightSeg0 = [new(w, 0), new(w, h)];
            _lightSeg1 = [new(0, h), new(w, h)];
        }

        #endregion

        #region Properties

        /// <summary>
        /// Gets or sets the text alignment inside the TextBox.
        /// </summary>
        [Category("Options")]
        public HorizontalAlignment TextAlign
        {
            get => _textAlign;
            set { _textAlign = value; TB?.TextAlign = value; }
        }
        private HorizontalAlignment _textAlign = HorizontalAlignment.Left;

        /// <summary>
        /// Gets or sets the maximum number of characters the user can enter.
        /// </summary>
        [Category("Options")]
        public int MaxLength
        {
            get => _maxLength;
            set { _maxLength = value; TB?.MaxLength = value; }
        }
        private int _maxLength = 32767;

        /// <summary>
        /// Gets or sets a value indicating whether the text is read-only.
        /// </summary>
        [Category("Options")]
        public bool ReadOnly
        {
            get => _readOnly;
            set { _readOnly = value; TB?.ReadOnly = value; }
        }
        private bool _readOnly = false;

        /// <summary>
        /// Gets or sets a value indicating whether text is displayed as the password character.
        /// </summary>
        [Category("Options")]
        public bool UseSystemPasswordChar
        {
            get => _useSystemPasswordChar;
            set { _useSystemPasswordChar = value; TB?.UseSystemPasswordChar = value; }
        }
        private bool _useSystemPasswordChar = false;

        /// <summary>
        /// Gets or sets a value indicating whether this is a multiline TextBox.
        /// </summary>
        [Category("Options")]
        public bool Multiline
        {
            get => _multiline;
            set
            {
                _multiline = value;
                if (TB != null)
                {
                    TB.Multiline = value;
                    UpdateInnerBounds();
                }
            }
        }
        private bool _multiline = false;

        /// <summary>
        /// Gets or sets the text displayed in the TextBox.
        /// </summary>
        [Category("Options")]
        public override string Text
        {
            get => base.Text;
            set
            {
                base.Text = value;
                if (TB != null && TB.Text != value) TB.Text = value;
            }
        }

        /// <summary>
        /// Gets or sets the font used by the TextBox.
        /// </summary>
        [Category("Options")]
        public override Font Font
        {
            get => base.Font;
            set
            {
                base.Font = value;
                if (TB != null)
                {
                    TB?.Font = value;
                    UpdateInnerBounds();
                }
            }
        }

        /// <summary>
        /// Gets or sets the background color. Also applied to the inner TextBox.
        /// </summary>
        public new Color BackColor
        {
            get => base.BackColor;
            set
            {
                if (base.BackColor == value) return;
                base.BackColor = value;
                TB?.BackColor = value;
                Invalidate();
            }
        }

        /// <summary>
        /// Gets or sets the foreground (text) color. Also applied to the inner TextBox.
        /// </summary>
        public new Color ForeColor
        {
            get => base.ForeColor;
            set
            {
                if (base.ForeColor == value) return;
                base.ForeColor = value;
                TB?.ForeColor = value;
                Invalidate();
            }
        }

        /// <summary>
        /// Gets or sets the inner top/left shadow color (second border layer).
        /// </summary>
        public Color ButtonShadow
        {
            get => _buttonShadow;
            set
            {
                if (_buttonShadow == value) return;
                _buttonShadow = value;
                ReplacePen(ref _penShadow, value);
                Invalidate();
            }
        }
        private Color _buttonShadow = SystemColors.ButtonShadow;

        /// <summary>
        /// Gets or sets the outer top/left dark shadow color (outermost border layer).
        /// </summary>
        public Color ButtonDkShadow
        {
            get => _buttonDkShadow;
            set
            {
                if (_buttonDkShadow == value) return;
                _buttonDkShadow = value;
                ReplacePen(ref _penDkShadow, value);
                Invalidate();
            }
        }
        private Color _buttonDkShadow = SystemColors.ControlDark;

        /// <summary>
        /// Gets or sets the inner bottom/right hilight color (second border layer).
        /// </summary>
        public Color ButtonHilight
        {
            get => _buttonHilight;
            set
            {
                if (_buttonHilight == value) return;
                _buttonHilight = value;
                ReplacePen(ref _penHilight, value);
                Invalidate();
            }
        }
        private Color _buttonHilight = SystemColors.ButtonHighlight;

        /// <summary>
        /// Gets or sets the outer bottom/right light color (outermost border layer).
        /// </summary>
        public Color ButtonLight
        {
            get => _buttonLight;
            set
            {
                if (_buttonLight == value) return;
                _buttonLight = value;
                ReplacePen(ref _penLight, value);
                Invalidate();
            }
        }
        private Color _buttonLight = SystemColors.ControlLight;

        #endregion

        #region Overrides — layout

        /// <summary>
        /// Adds the inner TextBox to the control's child collection when the handle is created.
        /// </summary>
        protected override void OnCreateControl()
        {
            base.OnCreateControl();
            if (!Controls.Contains(TB)) Controls.Add(TB);
        }

        /// <summary>
        /// Repositions the inner TextBox and rebuilds geometry when the control is resized.
        /// </summary>
        protected override void OnResize(EventArgs e)
        {
            base.OnResize(e);
            UpdateInnerBounds();
            RebuildGeometry();
        }

        /// <summary>
        /// Rebuilds geometry when the font changes.
        /// </summary>
        protected override void OnFontChanged(EventArgs e)
        {
            base.OnFontChanged(e);
            RebuildGeometry();
        }

        #endregion

        #region Overrides — focus and mouse

        /// <summary>
        /// Focuses the inner TextBox on mouse down.
        /// </summary>
        protected override void OnMouseDown(MouseEventArgs e)
        {
            base.OnMouseDown(e);
            TB?.Focus();
            Invalidate();
        }

        /// <summary>
        /// Focuses the inner TextBox on mouse up.
        /// </summary>
        protected override void OnMouseUp(MouseEventArgs e)
        {
            base.OnMouseUp(e);
            TB?.Focus();
            Invalidate();
        }

        /// <summary>
        /// Invalidates on mouse enter so the border can respond to hover if needed.
        /// </summary>
        protected override void OnMouseEnter(EventArgs e)
        {
            base.OnMouseEnter(e);
            Invalidate();
        }

        /// <summary>
        /// Invalidates on mouse leave.
        /// </summary>
        protected override void OnMouseLeave(EventArgs e)
        {
            base.OnMouseLeave(e);
            Invalidate();
        }

        private void TB_MouseDown(object sender, MouseEventArgs e) => Invalidate();
        private void TB_MouseEnter(object sender, EventArgs e) => Invalidate();
        private void TB_MouseLeave(object sender, EventArgs e) => Invalidate();
        private void TB_LostFocus(object sender, EventArgs e) => Invalidate();

        #endregion

        #region Overrides — text input

        /// <summary>
        /// Propagates the inner TextBox text to the outer control's Text property.
        /// Guards against a set loop by checking for equality first.
        /// </summary>
        private void OnInnerTextChanged(object sender, EventArgs e)
        {
            if (base.Text != TB?.Text) base.Text = TB?.Text;
        }

        /// <summary>
        /// Handles standard clipboard and undo keyboard shortcuts.
        /// </summary>
        private void OnInnerKeyDown(object sender, KeyEventArgs e)
        {
            if (!e.Control) return;

            switch (e.KeyCode)
            {
                case Keys.A: TB.SelectAll(); e.SuppressKeyPress = true; break;
                case Keys.C: TB.Copy(); e.SuppressKeyPress = true; break;
                case Keys.X: TB.Cut(); e.SuppressKeyPress = true; break;
                case Keys.V: TB.Paste(); e.SuppressKeyPress = true; break;
                case Keys.Z: TB.Undo(); e.SuppressKeyPress = true; break;
            }

            Invalidate();
        }

        #endregion

        #region Paint

        /// <summary>
        /// Paints the Win9x sunken 3D border directly onto the control surface.
        /// The inner TextBox renders its own text — this method draws only the border.
        /// </summary>
        protected override void OnPaint(PaintEventArgs e)
        {
            Graphics G = e.Graphics;
            G.SmoothingMode = SmoothingMode.HighSpeed;
            G.TextRenderingHint = DesignMode ? TextRenderingHint.ClearTypeGridFit : Program.Style.TextRenderingHint;

            // Background — fills the area not covered by the inner TextBox.
            G.Clear(BackColor);

            // Win9x sunken border — outermost to innermost, top-left then bottom-right.
            // DkShadow and Shadow on top+left give the pressed/inset appearance.
            if (_penDkShadow != null) { G.DrawLine(_penDkShadow, _dkShadowSeg0[0], _dkShadowSeg0[1]); G.DrawLine(_penDkShadow, _dkShadowSeg1[0], _dkShadowSeg1[1]); }
            if (_penShadow != null) { G.DrawLine(_penShadow, _shadowSeg0[0], _shadowSeg0[1]); G.DrawLine(_penShadow, _shadowSeg1[0], _shadowSeg1[1]); }
            if (_penHilight != null) { G.DrawLine(_penHilight, _hilightSeg0[0], _hilightSeg0[1]); G.DrawLine(_penHilight, _hilightSeg1[0], _hilightSeg1[1]); }
            if (_penLight != null) { G.DrawLine(_penLight, _lightSeg0[0], _lightSeg0[1]); G.DrawLine(_penLight, _lightSeg1[0], _lightSeg1[1]); }
        }

        #endregion

        #region Dispose

        /// <summary>
        /// Releases all cached GDI resources and the inner TextBox.
        /// </summary>
        protected override void Dispose(bool disposing)
        {
            if (disposing)
            {
                _penShadow?.Dispose();
                _penDkShadow?.Dispose();
                _penHilight?.Dispose();
                _penLight?.Dispose();
                _tb?.Dispose();
            }

            base.Dispose(disposing);
        }

        #endregion
    }
}